pkg/log/context.go

```diff
@@ -0,0 +1,184 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package log
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+type logKey struct{}
+
+// LevelKey is the context key for the log level
+type LevelKey struct{}
+
+var (
+	supportedFormatters = map[string]logrus.Formatter{
+		"json": &logrus.JSONFormatter{},
+		"text": &logrus.TextFormatter{},
+	}
+	mux          = sync.Mutex{}
+	once         = sync.Once{}
+	defaultEntry = logrus.NewEntry(logrus.StandardLogger())
+	// C is an alias for ForContext
+	C = ForContext
+	// R is an alias for ForContextProvider
+	R = ForContextProvider
+	// D is an alias for Default
+	D = Default
+)
+
+const (
+	// FieldComponentName is the key of the component field in the log message.
+	FieldComponentName = "component"
+	// FieldCorrelationID is the key of the correlation id field in the log message.
+	FieldCorrelationID = "correlation_id"
+)
+
+// Contexter interface
+type Contexter interface {
+	Context() context.Context
+}
+
+// Settings type to be loaded from the environment
+type Settings struct {
+	Level  string
+	Format string
+}
+
+// DefaultSettings returns default values for Log settings
+func DefaultSettings() *Settings {
+	return &Settings{
+		Level:  "debug",
+		Format: "text",
+	}
+}
+
+// Validate validates the logging settings
+func (s *Settings) Validate() error {
+	if len(s.Level) == 0 {
+		return fmt.Errorf("validate Settings: LogLevel missing")
+	}
+	if len(s.Format) == 0 {
+		return fmt.Errorf("validate Settings: LogFormat missing")
+	}
+	return nil
+}
+
+// Configure creates a new context with a logger using the provided settings.
+func Configure(ctx context.Context, settings *Settings) context.Context {
+	once.Do(func() {
+		level, err := logrus.ParseLevel(settings.Level)
+		if err != nil {
+			panic(fmt.Sprintf("Could not parse log level configuration: %s", err))
+		}
+		formatter, ok := supportedFormatters[settings.Format]
+		if !ok {
+			panic(fmt.Sprintf("Invalid log format: %s", settings.Format))
+		}
+		logrus.SetLevel(level)
+		logrus.SetFormatter(formatter)
+		logger := &logrus.Logger{
+			Formatter: formatter,
+			Level:     level,
+			Out:       os.Stdout,
+			Hooks:     make(logrus.LevelHooks),
+		}
+		defaultEntry = logrus.NewEntry(logger)
+	})
+	return ContextWithLogger(ctx, defaultEntry)
+}
+
+// ForContext retrieves the current logger from the context, configured for the provided component.
+// Optionally keys mapped to values from the context can be provided.
+// If no logger is present in the context, the default logger is returned
+func ForContext(ctx context.Context, component string, keys ...interface{}) *logrus.Entry {
```

Comment:
we might store the loggers created with keys the packages (log sources) - this way we won't have the ovehead of reflection to determine the log source every time we log.
Also currently we have the overhead that if we something in 20 packages in one request we will create 20 loggers for that - actually i think currently we create a new logger per log message.

We might attempt to store an array of base (package loggers) that have some fields prepolulated - such as logSource (e.x. /api/controllers/osb.go). Then as the request flows through certain filters can reuse the base loggers and attach additional fields (they obtain a copy of the base logger and let's say attach a correlationid field)

It would prolly make sense that such fields (like correlation id) are attached in seperate filters)